src/sys/petsc_interface.cxx

```diff
@@ -0,0 +1,173 @@
+ *
+ **************************************************************************/
+#include "mpi.h"
+#include <memory>
+#include <vector>
+
+#include <bout/mesh.hxx>
+#include <bout/paralleltransform.hxx>
+#include <bout/petsc_interface.hxx>
+#include <bout/petsclib.hxx>
+#include <bout/region.hxx>
+#include <bout_types.hxx>
+#include <boutcomm.hxx>
+
+#ifdef BOUT_HAS_PETSC
+
+// GlobalIndexer implementation
+
+bool GlobalIndexer::initialisedGlobal = false;
+IndexerPtr GlobalIndexer::globalInstance;
+Mesh* GlobalIndexer::globalmesh = nullptr;
+
+IndexerPtr GlobalIndexer::getInstance(Mesh* localmesh) {
+  // Check that the identity of bout::globals::mesh hasn't changed
+  // since last call. (Needed for unit tests)
+  if (bout::globals::mesh != globalmesh) {
+    globalmesh = bout::globals::mesh;
+    initialisedGlobal = false;
+  }
+  if (localmesh == globalmesh) {
+    if (!initialisedGlobal) {
+      globalInstance = std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+      initialisedGlobal = true;
+    }
+    return globalInstance;
+  } else {
+    return std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+  }
+}
+
+void GlobalIndexer::initialiseTest() {
+  registerFieldForTest(indices3D);
+  registerFieldForTest(indices2D);
+  registerFieldForTest(indicesPerp);
+}
+
+void GlobalIndexer::initialise() {
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+  // Communicate a second time to get any corner values
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+}
+
+Mesh* GlobalIndexer::getMesh() { return fieldmesh; }
+
+PetscInt GlobalIndexer::getGlobal(Ind2D ind) {
+  return static_cast<PetscInt>(indices2D[ind] + 0.5);
+}
+
+PetscInt GlobalIndexer::getGlobal(Ind3D ind) {
+  return static_cast<PetscInt>(indices3D[ind] + 0.5);
+}
+
+PetscInt GlobalIndexer::getGlobal(IndPerp ind) {
+  return static_cast<PetscInt>(indicesPerp[ind] + 0.5);
+}
+
+void GlobalIndexer::registerFieldForTest(FieldData& UNUSED(f)) {
+  // This is a place-holder which does nothing. It can be overridden
+  // by descendent classes if necessary to set up testing.
+  return;
+}
+
+void GlobalIndexer::registerFieldForTest(FieldPerp& UNUSED(f)) {
+  // This is a place-holder which does nothing. It can be overridden
+  // by descendent classes if necessary to set up testing.
+  return;
+}
+
+GlobalIndexer::GlobalIndexer(Mesh* localmesh)
+    : fieldmesh(localmesh), indices3D(-2., localmesh), indices2D(-2., localmesh),
+      indicesPerp(-2., localmesh) {
+  // Set up the 3D indices
+  int counter = localmesh->globalStartIndex3D();
+  for (RangeIterator it = localmesh->iterateBndryLowerY(); !it.isDone(); it++) {
+    for (int z = 0; z < localmesh->LocalNz; z++) {
+      if (it.ind == localmesh->xstart)
+        indices3D(it.ind - 1, localmesh->ystart - 1, z) = counter++;
+      if (it.ind == localmesh->xend)
+        indices3D(it.ind + 1, localmesh->ystart - 1, z) = counter++;
```

Comment:
In principle this doesn't matter but I agree it's messy. However, since I refactored to use new Region objects this is a moot point. The regions will have everything in the obvious order.